### PR TITLE
Fix reëntrant events/spans escaping per-layer filters

### PR DESCRIPTION
**Pull request by maddiemort**
This is a fork-internal PR, intended to be a more rough-draft version of the change that will eventually be filed upstream. This incorporates the per-layer filter escape fix into the `ditto` branch, so it can be used in Ditto code.

Please see getditto/ditto#13306 for the motivation, and tokio-rs/tracing#2704 for context & discussion.

The fix consists of two primary changes:

- Instead of per-layer filters simply storing one instance of `FilterMap` inside their thread-local `FilterState`, `FilterState` now holds a "stack" of filter maps, and per-layer filters push the current state to the stack when they discover that they're about to write to a state that's stale
    - The determination of whether a state is stale is made by the addition of a second bitmap, `seen`, which will contain a bit at each filter's ID if it's already gone through a filtering pass on the current state. That means that it must be inside a reëntrant call (because the filter state should have been cleared before it's asked to evaluate its enabledness again, if the call is not reëntrant).
- `Filtered` layers, which are layers that have a per-layer filter attached, no longer ever store their interest as "always", instead knocking down an "always" interest to "sometimes". This likely represents a performance regression, but it's a necessary one in order to force the reëvaluation of `enabled()` for every span/event - if we use "always" and short-circuit that call, we see incorrect behavior for some kinds of reëntrant event scenarios (such as the new test `enabled_event_inside_event`)

### CODE REVIEW

**Comment by maddiemort** [comment]
These clippy failures come from the base branch 😕

**Review by hamchapman — APPROVED** [review]
Superb work 👏 

The documentation is particularly top notch 👌

**Comment by maddiemort** [comment]
Thank you! I'm going to go ahead and merge this (to the `ditto` branch). Any reviews after the fact are appreciated, and can be addressed later 😄

File: tracing-subscriber/src/filter/layer_filters/mod.rs [hunk 1]
```diff
@@ -143,6 +150,48 @@ pub(crate) struct FilterState {
...
+        let nested = FilterMapStack::default();
+
+        self.current.swap(&nested.current);
+        // Any filters already nested in `self` represent an existing stack, and
+        // so should now be nested under the new top of the stack.
+        self.nested.swap(&nested.nested);
+
+        // Our state has been moved into `nested`, leaving us with fresh values
+        // for our fields, so nest `nested` under `self`.
+        self.nested.replace(Some(Box::new(nested)));
```
Comment (danielhenrymantilla) [inline tracing-subscriber/src/filter/layer_filters/mod.rs:176]: Micro-opt: you could probably "lock" once here.

File: tracing-subscriber/src/filter/layer_filters/mod.rs [hunk 2]
```diff
@@ -143,6 +150,48 @@ pub(crate) struct FilterState {
...
+    fn pop_or_clear(&self) {
+        // Like in `push()`, we can't move `self`. Instead, let's just overwrite
+        // the values of our fields by swapping them out from the fields of
+        // `self.nested`.
+        let to_swap = self.nested.take().unwrap_or_default();
+
+        self.current.swap(&to_swap.current);
+        // If there were additional nested filter maps under `self.nested`,
+        // they'll now become nested under `self`.
+        self.nested.swap(&to_swap.nested);
```
Comment (danielhenrymantilla) [inline tracing-subscriber/src/filter/layer_filters/mod.rs:191]: Similarly, you could lock once `self.nested` as well

**Review by danielhenrymantilla — COMMENTED** [review]
Nice, I like the `.bits` renamed as `.disabled` for clarity 🙏

**Review by danielhenrymantilla — COMMENTED** [review]
Actually, rather than using larger `.borrow_mut()`s on `nested`, we ought to be able to just use a plain `Cell` as well? (I haven't been able to see any `.borrow{,_mut}()`s)

**Pull request merged**